README.md
# QAFactEval: Improved QA-Based Factual Consistency Evaluation for Summarization

This is the official code repository for the NAACL 2022 paper [QAFactEval: Improved QA-Based Factual Consistency Evaluation for Summarization](https://arxiv.org/abs/2112.08542)
by [Alexander R. Fabbri](https://twitter.com/alexfabbri4), [Chien-Sheng Wu](https://twitter.com/jasonwu0731), [Wenhao Liu](https://twitter.com/owenhaoliu), and [Caiming Xiong](https://twitter.com/caimingxiong). 

In our paper, we conduct an extensive comparison of the components of QA-based metrics for factual consistency evaluation in summarization. Our optimized metric builds on [QAEval](https://direct.mit.edu/tacl/article/doi/10.1162/tacl_a_00397/106792/Towards-Question-Answering-as-an-Automatic-Metric) with question consistency filtering and an improved answer overlap metric, leading to a 14% average improvement over previous QA-based metrics on the [SummaC](https://direct.mit.edu/tacl/article/doi/10.1162/tacl_a_00453/109470/SummaC-Re-Visiting-NLI-based-Models-for) factual consistency benchmark. 

## Table of Contents

1. [Updates](#updates)
2. [Using QAFactEval](#using-qafacteval)
3. [Citation](#citation)
4. [License](#license)


## Updates
_5/2/2022_ - Initial commit! :) 

## Using QAFactEval

You can install qafacteval via pip:
```bash
pip install qafacteval
```

You can also install from source:

```bash
git clone https://github.com/salesforce/QAFactEval
cd QAFactEval
pip install -e .
```

### For use in scripts
Download the required pretrained models using `download_models.sh`.

See `run.py` for an example of using the QAFactEval metric:

```python
from qafacteval import QAFactEval
kwargs = {"cuda_device": 0, "use_lerc_quip": True, \
        "verbose": True, "generation_batch_size": 32, \
        "answering_batch_size": 32, "lerc_batch_size": 8}

model_folder = "" # path to models downloaded with download_models.sh
metric = QAFactEval(
    lerc_quip_path=f"{model_folder}/quip-512-mocha",
    generation_model_path=f"{model_folder}/generation/model.tar.gz",
    answering_model_dir=f"{model_folder}/answering",
    lerc_model_path=f"{model_folder}/lerc/model.tar.gz",
    lerc_pretrained_model_path=f"{model_folder}/lerc/pretraining.tar.gz",
    **kwargs
)

results = metric.score_batch(["This is a source document"], [["This is a summary."]], return_qa_pairs=True)
score = results[0][0]['qa-eval']['lerc_quip']

```

## Citation

When referencing this repository, please cite [this paper](https://arxiv.org/abs/2112.08542):

```bibtex
@misc{fabbri-etal-2022-qafacteval,
  title = {QAFactEval: Improved QA-Based Factual Consistency Evaluation for Summarization},
  author = {Alexander R. Fabbri and Chien-Sheng Wu and Wenhao Liu and Caiming Xiong},
  year={2022},
  eprint={2112.08542},
  archivePrefix={arXiv},
  primaryClass={cs.CL},
  url = {https://arxiv.org/abs/2112.08542},
}
```

## License

This repository is released under the [BSD-3 License](LICENSE.txt).


qaeval/changes
Modified https://github.com/danieldeutsch/qaeval/blob/master/qaeval/metric.py#L351 
Added LERC (QuIP) download to https://raw.githubusercontent.com/danieldeutsch/repro/master/models/deutsch2021/scripts/download-models.sh


lerc_quip.py
"""
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
"""

from typing import Dict, List, Set

from qaeval.scoring.scorers.scorer import Scorer
from transformers import AutoModelForSequenceClassification, AutoTokenizer

def predict(our_model, our_tokenizer, batch_sentences, device):
    inputs = our_tokenizer(batch_sentences, max_length=512,  truncation=True, \
        padding="max_length", return_tensors="pt")
    outputs = our_model(input_ids=inputs["input_ids"].to(device), \
        attention_mask=inputs["attention_mask"].to(device))
    outputs = [x[0] for x in outputs[0].cpu().tolist()]
    outputs = [{"pred_score": x} for x in outputs]

    return outputs


class LERCQuipScorer(Scorer):
    def __init__(self, lerc_quip_path: str, cuda_device: int, batch_size: int = 8) -> None:
        self.device = cuda_device

        self.predictor = AutoModelForSequenceClassification.from_pretrained(lerc_quip_path).to(self.device)
        self.predictor.eval()

        self.tokenizer = AutoTokenizer.from_pretrained(lerc_quip_path)
        self.batch_size = batch_size

    def keys(self) -> Set[str]:
        return {'lerc_quip'}

    def _score_single_ref(
        self,
        context: str,
        questions: List[str],
        answers: List[str],
        predictions: List[str],
        probabilities: List[float],
        null_probabilities: List[float]
    ) -> List[Dict[str, float]]:
        input_dicts = []
        indices = []
        for i, (answer, question, prediction, probability, null_probability) in \
                enumerate(zip(answers, questions, predictions,
                    probabilities, null_probabilities)):
            if probability > null_probability:
                sentence1 = f"{question} <q> {answer} <r> {prediction} <c> {context}"
                input_dicts.append(sentence1)
                indices.append(i)

        output_dicts = []
        for i in range(0, len(input_dicts), self.batch_size):
            batch = input_dicts[i:i + self.batch_size]
            output_dicts.extend(predict(self.predictor, self.tokenizer, batch, self.device))
        assert len(output_dicts) == len(input_dicts)

        scores = [0.0] * len(questions)
        for i, output_dict in zip(indices, output_dicts):
            scores[i] = output_dict['pred_score']
        scores = [{'lerc_quip': s} for s in scores]
        return scores


qafacteval.py
"""
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
"""

from typing import Dict, List, Union
from transformers.data.metrics.squad_metrics import compute_f1

from qaeval import QAEval
from lerc_quip import LERCQuipScorer

MetricsDict = Dict[str, float]
SummaryType = Union[str, List[str]]

def get_filter(qa_summ, answers_ref):
    answerable = []
    a_orig = []
    for qa_summ_ in qa_summ:
        answerability = (qa_summ_[1] > qa_summ_[2])
        answerable.append(answerability)
        a_orig.append(qa_summ_[0])

    f1s = [compute_f1(answer, prediction) for \
        answer, prediction in zip(answers_ref, a_orig)]
    bool_f1 = [x > 0.60 for x in f1s]
    bool_total = [x and y for x, y in zip(bool_f1, answerable)]
    return bool_total

class QAFactEval(QAEval):
    def __init__(
            self,
            lerc_quip_path: str,
            use_lerc_quip: bool,
            lerc_batch_size: int,
            cuda_device: int,
            *args,
            **kwargs):
        try:
            import spacy
            spacy.load("en_core_web_sm")
        except:
            import spacy.cli
            spacy.cli.download("en_core_web_sm")
        super().__init__(*args, **kwargs)

        if use_lerc_quip:
            lerc_quip = LERCQuipScorer(lerc_quip_path=lerc_quip_path, \
                cuda_device=cuda_device, batch_size=lerc_batch_size)
            self.scorer.scorers.append(lerc_quip)

    def score_batch_qafacteval(
        self,
        source: List[SummaryType],
        summaries: List[List[SummaryType]],
        qa_pairs_precomputed: List = None,
        predictions_lists: List = None,
        return_qa_pairs: bool = False,
            ) -> List[List[MetricsDict]]:
        source = self._flatten_summaries(source)
        summaries = self._flatten_references_list(summaries)

        # Remove any input source that are empty. They mess up the processing otherwise
        (
            source,
            summaries,
            is_empty_list,
        ) = self._get_empty_summary_mask(source, summaries)


        if qa_pairs_precomputed:
            qa_pairs_lists = qa_pairs_precomputed
        else:
            qa_pairs_lists = self._generate_qa_pairs(summaries)

        # question_consistency
        # TODO: only uses one summary here
        summaries_cons = [x[0] for x in summaries]
        predictions_lists_consistency = self._answer_questions(summaries_cons, qa_pairs_lists)

        qa_pairs_lists_cons = []
        for x, cur_qa_pair in zip(predictions_lists_consistency, qa_pairs_lists):
            qa_summ_new = [[x["prediction"], x["probability"], \
                x["null_probability"]] for x in x[0]]
            answers_ref = [x["answer"] for x in cur_qa_pair[0]]

            bool_total = get_filter(qa_summ_new, answers_ref)

            cur_qa_pair_keep = [x for count, x in enumerate(cur_qa_pair[0]) if bool_total[count]]
            if not cur_qa_pair_keep:
                cur_qa_pair_keep = []
            qa_pairs_lists_cons.append([cur_qa_pair_keep])

        if predictions_lists:
            predictions_lists = predictions_lists
        else:
            predictions_lists = self._answer_questions(source, qa_pairs_lists_cons)

        metrics_list, scores_lists = self._score_predictions(
            source, qa_pairs_lists_cons, predictions_lists
        )

        if return_qa_pairs:
            output = self._combine_outputs(
                metrics_list, qa_pairs_lists_cons, predictions_lists, scores_lists
            )
        else:
            output = metrics_list

        output = self._insert_empty_outputs(output, is_empty_list, return_qa_pairs)
        output_final = []
        # Add consistency info for analysis
        for out, qa_pairs_list, predictions_cons in \
                zip(output, qa_pairs_lists, predictions_lists_consistency):
            output_final.append((out[0], out[1], qa_pairs_list, predictions_cons))
        return output_final


run.py
"""
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
"""

import json
import argparse
from qafacteval import QAFactEval


if __name__ == '__main__':
    parser = argparse.ArgumentParser('Running QAFactEval')
    parser.add_argument('--fname', default=None, required=True)
    parser.add_argument('--outfname', default=None, required=True)
    parser.add_argument('--model_folder', default=None, required=True)
    parser.add_argument('--cuda_device', default=0, type=int)
    parser.add_argument('--use_lerc_quip', default=False, action='store_true')
    parser.add_argument('--verbose', default=False, action='store_true')
    parser.add_argument('--generation_batch_size', default=32, type=int)
    parser.add_argument('--answering_batch_size', default=32, type=int)
    parser.add_argument('--lerc_batch_size', default=8, type=int)

    args = parser.parse_args()
    kwargs = {"cuda_device": args.cuda_device, "use_lerc_quip": args.use_lerc_quip, \
        "verbose": args.verbose, "generation_batch_size": args.generation_batch_size, \
        "answering_batch_size": args.answering_batch_size, "lerc_batch_size": args.lerc_batch_size}

    metric = QAFactEval(
        lerc_quip_path=f"{args.model_folder}/quip-512-mocha",
        generation_model_path=f"{args.model_folder}/generation/model.tar.gz",
        answering_model_dir=f"{args.model_folder}/answering",
        lerc_model_path=f"{args.model_folder}/lerc/model.tar.gz",
        lerc_pretrained_model_path=f"{args.model_folder}/lerc/pretraining.tar.gz",
        **kwargs
    )

    # results = metric.score_batch(["This is a source document"], \
    #   [["This is a summary."]], return_qa_pairs=True)
    # print(results[0][0]['qa-eval']['lerc_quip'])

    candidates = []
    references_list = []
    datas = []
    with open(args.fname) as f:
        for line in f:
            data = json.loads(line)
            datas.append(data)
            doc = data['document']['text']
            summ = data['claim']
            candidates.append(doc)
            references_list.append([summ])

    results = metric.score_batch_qafacteval(candidates, references_list, return_qa_pairs=True)
    with open(args.outfname, "w") as out:
        for (metrics, qa_pairs, qa_pairs_original, predictions_cons), data in zip(results, datas):
            data["metrics"] = metrics
            data["qa_pairs"] = qa_pairs
            data["qa_pairs_nonfiltered"] = qa_pairs_original
            data["qa_summary"] = predictions_cons
            out.write(json.dumps(data) + "\n")
